langkit/templates/pkg_implementation_body_ada.mako

```diff
@@ -1336,60 +1364,56 @@ package body ${ada_lib_name}.Implementation is
          --
          --  This reasoning applies to environments that belong to foreign
          --  units, but also to the root environment.
-         Is_Foreign (Dest_Env, Self) and then Is_Synthetic (Mapping.Val)
+         Is_Foreign (Dest_Env, Self) and then Is_Synthetic (Value)
       then
          raise Property_Error with
             "Cannot add a synthetic node to a lexical env from another"
             & " analysis unit";
-      end if;
 
-      --  If requested, reject foreign destination environments.  Note that
-      --  this detects only explicit destination environments
-      --  (Mapping.Dest_Env): Set_Initial_Env already sanitizes initial
-      --  environments (State.Current_Env).
-      --
-      --  This is an attempt at identifying uses of the unsound relocation
-      --  mechanism (as opposed to named environments), so this applies to all
-      --  foreign environments (root scope included).
-      if DSL_Location'Length > 0
-         and then Is_Foreign_Strict (Mapping.Dest_Env, Self)
+      elsif
+         --  If requested, reject foreign destination environments.  Note that
+         --  this checks only explicit destination environments
+         --  (Dest_Env_Fallback): Set_Initial_Env already sanitized initial
+         --  environments (State.Current_Env).
+         --
+         --  This is an attempt at identifying uses of the unsound relocation
+         --  mechanism (as opposed to named environments), so this applies to
+         --  all foreign environments (root scope included).
+         DSL_Location'Length > 0
```

Comment:
I would be in favor of adding another condition: `and then Dest_NED = null`, so that this check is only done if we actually did go through the fallback. This allows reusing the old code (before the named envs) as the fallback code to handle the non-named cases.

Reply:
As discussed live: when there is an env name, we don’t evaluate the fallback env, so `Dest_Env_Fallback` is `Empty_Env`, and thus the check is not triggered in that case. So the extra condition you suggest is redundant. I’ve added a comment to explicitly state this. Thanks!